libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TriggerConditions/OnCondition.cs

```diff
@@ -278,6 +246,79 @@ public virtual IEnumerable<Dialog> GetDependencies()
             yield return this.ActionScope;
         }
 
+        /// <summary>
+        /// Create the expression for this condition.
+        /// </summary>
+        /// <remarks>
+        /// Override this in base classes to create the expression for this trigger.
+        /// </remarks>
+        /// <returns>Expression used to evaluate this rule. </returns>
+        protected virtual Expression CreateExpression()
+        {
+            var allExpressions = new List<Expression>();
+
+            if (this.Condition != null)
+            {
+                allExpressions.Add(this.Condition.ToExpression());
+            }
+
+            if (this.extraConstraints.Any())
+            {
+                allExpressions.AddRange(this.extraConstraints);
```

Comment:
Locks are acquired around this.extraConstraints within the rest of this class.  If this `protected virtual` method does not acquire the lock, and does not enable overridden versions to do the same: i don't see how this is threadsafe.

Reply:
it's thread safe because the lock is for writing to the private member variable, which this method does NOT do.,  
Look at GetExpression(), That has lock for purposes of assiognemnt to this._fullConstraint, not for the method call itself.

---
In reply to: [520151804](https://github.com/microsoft/botbuilder-dotnet/pull/4935#discussion_r520151804) [](ancestors = 520151804)

Reply:
@tomlm  maybe it is unlikely someone will run into this, but a caller of this method is not acquiring a lock for this.extraConstraints and this list could presumably be added to from other threads by calling these public methods:

```cs
public void AddExternalCondition(Expression condition)
public void AddExternalCondition(string condition)
```

If this class is intended to be threadsafe, a lock should be acquired while reading this list.  A List is only safe to read from if there is a guarantee no other threads are modifying it: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?redirectedfrom=MSDN&view=net-5.0#thread-safety

`It is safe to perform multiple read operations on a List<T>, but issues can occur if the collection is modified while it's being read. To ensure thread safety, lock the collection during a read or write operation.`